crates/topos-p2p/src/behaviour/grpc/stream.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    io,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use futures::{AsyncRead as FuturesAsyncRead, AsyncWrite as FuturesAsyncWrite, Future};
+use http::Uri;
+use libp2p::{swarm::ConnectionId, PeerId};
+use pin_project::pin_project;
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    sync::mpsc,
+};
+use tonic::transport::{server::Connected, Channel, Endpoint};
+use tower::{BoxError, Service};
+use tracing::info;
+
+/// Manage a gRPC Stream linked to an open [`libp2p::Stream`]
+#[pin_project]
+pub(crate) struct GrpcStream {
+    #[pin]
+    stream: libp2p::Stream,
+    peer_id: PeerId,
+    connection_id: libp2p::swarm::ConnectionId,
+}
+
+/// Outbound GrpcStream initialization struct
+#[pin_project]
+struct InitializedGrpcOutboundStream {
+    #[pin]
+    stream_rx: Arc<Mutex<mpsc::Receiver<Result<GrpcStream, BoxError>>>>,
+}
+
+/// Fully negotiated Outbound GrpcStream
+#[pin_project]
+struct NegotiatedGrpcOutboundStream {
+    #[pin]
+    stream_rx: Arc<Mutex<mpsc::Receiver<Result<GrpcStream, BoxError>>>>,
+}
+
+impl Future for NegotiatedGrpcOutboundStream {
+    type Output = Result<GrpcStream, BoxError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let fut = self.project();
+        let x = fut
+            .stream_rx
+            .lock()
+            .unwrap()
```

Comment:
I know of the stdlib's mutexe poisoning stuff, but I think we should use `expect("why it's reasonable to make the assumptions we make")` as a pattern in all our code. It amounts to the same thing, just a wee bit clearer to the reader that we know what we're doing?